deploy/stacks/bootstrap_stack.py

```diff
@@ -0,0 +1,201 @@
+from aws_cdk import Duration
+from aws_cdk import (
+    aws_iam as iam,
+    aws_ec2 as ec2,
+    aws_lambda as _lambda,
+)
+from aws_cdk.triggers import TriggerFunction
+
+from .pyNestedStack import pyNestedClass
+
+
+class BootstrapStack(pyNestedClass):
+    def __init__(
+        self,
+        scope,
+        id,
+        handler,
+        envname='dev',
+        resource_prefix='dataall',
+        vpc=None,
+        vpce_connection=None,
+        ecr_repository=None,
+        image_tag=None,
+        pivot_role_name=None,
+        connections: [ec2.Connections] = [],
+        **kwargs,
+    ):
+        super().__init__(scope, id, **kwargs)
+
+        if self.node.try_get_context('image_tag'):
+            image_tag = self.node.try_get_context('image_tag')
+        image_tag = f'lambdas-{image_tag}'
+
+        env = {'envname': envname, 'LOG_LEVEL': 'INFO'}
+
+        self.trigger_function = TriggerFunction(
+            self,
+            f'BootstrapFunction-{handler}',
+            function_name=f'{resource_prefix}-{envname}-{handler.replace(".", "_")}',
+            description=f'dataall {handler} function',
+            role=self.create_function_role(envname, resource_prefix, handler, pivot_role_name),
+            code=_lambda.Code.from_ecr_image(repository=ecr_repository, tag=image_tag, cmd=[handler]),
+            vpc=vpc,
+            security_groups=[self.create_lambda_sgs(envname, handler, resource_prefix, vpc)],
+            memory_size=256,
+            timeout=Duration.minutes(15),
+            environment=env,
+            tracing=_lambda.Tracing.ACTIVE,
+            runtime=_lambda.Runtime.FROM_IMAGE,
+            handler=_lambda.Handler.FROM_IMAGE,
+        )
+
+        if vpce_connection:
+            self.trigger_function.connections.allow_from(
+                vpce_connection,
+                ec2.Port.tcp_range(start_port=1024, end_port=65535),
+                'Allow Lambda from VPC Endpoint',
+            )
+            self.trigger_function.connections.allow_to(
+                vpce_connection, ec2.Port.tcp(443), 'Allow Lambda to VPC Endpoint'
+            )
+
+        self.trigger_function.connections.allow_to(
+            ec2.Peer.any_ipv4(), ec2.Port.tcp(443), 'Allow NAT Internet Access SG Egress'
+        )
+
+        for connection in connections:
+            connection.allow_default_port_from(
+                self.trigger_function, f'Allow dataall {self.trigger_function.function_name}'
+            )
+
```

Comment:
We can scope down the permissions of the role of the `save_permissions`, it needs fewer permissions than the generic `api_handler` role. Especially important is to restrict the permissions to other IAM roles, e.g. we do not need to assume the pivot role or the cdk exec role in other accounts. Similar with Pass role, we should aim to have least-minimum-permissions in the save_permissions_role. I think we barely need a couple of permissions, I only identified parameter store and secrets manager permissions to get_engine. Maybe some KMS permissions for the encryption + some logs, xray

Reply:
I guess this comment goes to `create_function_role`, right? and yes you are absolutely right, I copy/pasted the perms from the api_handler and forgot to checkout the trimmed down version. Will update the commit.

Reply:
Resolved now!